source/common/config/watch_map.h

```diff
@@ -0,0 +1,118 @@
+// The xDS machinery's "watch" concept accomplishes that, while avoiding parallel redundant xDS
+// requests for X. Each of those subscriptions is viewed as a "watch" on X, while behind the scenes
+// there is just a single real subscription to that resource name.
+//
+// This class maintains the watches<-->subscription mapping: it
+// 1) delivers updates to all interested watches, and
+// 2) tracks which resource names should be {added to,removed from} the subscription when the
+//    {first,last} watch on a resource name is {added,removed}.
+//
+// #1 is accomplished by WatchMap's implementation of the SubscriptionCallbacks interface.
+// This interface allows the xDS client to just throw each xDS update message it receives directly
+// into WatchMap::onConfigUpdate, rather than having to track the various watches' callbacks.
+//
+// The information for #2 is returned by updateWatchInterest(); the caller should use it to
+// update the subscription accordingly.
+//
+// A WatchMap is assumed to be dedicated to a single type_url type of resource (EDS, CDS, etc).
+class WatchMap : public SubscriptionCallbacks, public Logger::Loggable<Logger::Id::config> {
+public:
+  WatchMap() {}
+
+  // An opaque token given out to users of WatchMap, to identify a given watch.
+  using Token = uint64_t;
+
+  // Adds 'callbacks' to the WatchMap, with no resource names being watched.
+  // (Use updateWatchInterest() to add some names).
+  // Returns a new token identifying the newly added watch.
+  Token addWatch(SubscriptionCallbacks& callbacks);
+
+  // Returns true if this was the very last watch in the map.
+  // Expects that the watch to be removed has already had all of its resource names removed via
+  // updateWatchInterest().
+  bool removeWatch(Token token);
+
+  // Updates the set of resource names that the given watch should watch.
+  // Returns any resource name additions/removals that are unique across all watches. That is:
+  // 1) if 'resources' contains X and no other watch cares about X, X will be in pair.first.
+  // 2) if 'resources' does not contain Y, and this watch was the only one that cared about Y,
+  //    Y will be in pair.second.
+  std::pair<std::set<std::string>, std::set<std::string>>
+  updateWatchInterest(Token token, const std::set<std::string>& update_to_these_names);
+
+  // SubscriptionCallbacks
+  virtual void onConfigUpdate(const Protobuf::RepeatedPtrField<ProtobufWkt::Any>& resources,
+                              const std::string& version_info) override;
+  virtual void
+  onConfigUpdate(const Protobuf::RepeatedPtrField<envoy::api::v2::Resource>& added_resources,
+                 const Protobuf::RepeatedPtrField<std::string>& removed_resources,
+                 const std::string& system_version_info) override;
+
+  virtual void onConfigUpdateFailed(const EnvoyException* e) override;
+
+  virtual std::string resourceName(const ProtobufWkt::Any&) override {
+    NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
```

Comment:
Is this where putting `resourceName` on subscription callbacks breaks down? Recalling out earlier conversation..

Reply:
Yes, although to be fair to `resourceName`, this is kind of an off-label implementation of SubscriptionCallbacks. But yeah, IMO requiring (someday) the xDS protocol to attach names to the opaque resource blobs is a better approach.